Whiskey/Tasks/Invoke-WhiskeyParallelTask.ps1

```diff
@@ -0,0 +1,118 @@
+                        Where-Object { $ExcludeProperty -notcontains $_.Name } |
+                        Where-Object { $_.GetSetMethod($false) } |
+                        Select-Object -ExpandProperty 'Name' |
+                        ForEach-Object { Write-Debug ('{0}  {1} -> {2}' -f $_,$Destination.$_,$Source.$_) ; $Destination.$_ = $Source.$_ }
+
+                }
+
+                $VerbosePreference = $using:VerbosePreference
+                $DebugPreferece = $using:DebugPreference
+                $whiskeyModulePath = $using:whiskeyModulePath 
+                $originalContext = $using:TaskContext
+                $taskName = $using:rsTaskName
+                $taskParameter = $using:rsTaskParameter
+
+                Import-Module -Name $whiskeyModulePath
+                $moduleRoot = $whiskeyModulePath | Split-Path
+
+                . (Join-Path -Path $moduleRoot -ChildPath 'Functions\Use-CallerPreference.ps1' -Resolve)
+                . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyContextObject.ps1' -Resolve)
+                . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyBuildMetadataObject.ps1' -Resolve)
+                . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyVersionObject.ps1' -Resolve)
+
+                $buildInfo = New-WhiskeyBuildMetadataObject
+                Sync-ObjectProperty -Source $originalContext.BuildMetada -Destination $buildInfo -Exclude @( 'BuildServer' )
+                if( $originalContext.BuildMetadata.BuildServer )
+                {
+                    $buildInfo.BuildServer = $originalContext.BuildMetadata.BuildServer
+                }
+                
+                $buildVersion = New-WhiskeyVersionObject
+                Sync-ObjectProperty -Source $originalContext.Version -Destination $buildVersion -ExcludeProperty @( 'SemVer1', 'SemVer2', 'SemVer2NoBuildMetadata' )
+                $buildVersion.SemVer1 = $originalContext.Version.SemVer1.ToString()
+                $buildVersion.SemVer2 = $originalContext.Version.SemVer2.ToString()
+                $buildVersion.SemVer2NoBuildMetadata = $originalContext.Version.SemVer2NoBuildMetadata.ToString()
+
+                $context = New-WhiskeyContextObject
+                Sync-ObjectProperty -Source $originalContext -Destination $context -ExcludeProperty @( 'BuildMetadata', 'Version' )
+
+                $context.BuildMetadata = $buildInfo
+                $context.Version = $buildVersion
+
+                Invoke-WhiskeyTask -TaskContext $context -Name $taskName -Parameter $taskParameter
+            }
+            $job | Add-Member -MemberType NoteProperty -Name 'TaskIndex' -Value $taskIdx
+            [void]$jobs.Add($job)
+    }
+
+    try
+    {
+        foreach( $job in $jobs )
+        {
+            Write-WhiskeyVerbose -Context $TaskContext -Message ('[{0}][{1}]  Waiting for background task.' -f $job.TaskIndex,$job.Name)
+            $job | Wait-Job | Receive-Job

```

Comment:
This line is synchronous right? Doesn't this mean we're only returning the job output in-order, as we're waiting for each task to complete in the order they were started? Meaning job 4's output won't be returned until we've returned job 1, 2, and 3, even though job 4 might have finished first? The documentation states that we're returning the output **"as each task finishes"** and not after all the tasks have finished in-order.
Also, what happens if job 2 finishes but job 1 fails? Do we get the output of job 2 before we stop Whiskey due to job 1's failure or is job 2's output simply lost when we fail the build?

Reply:
It means we wait until the first task finishes before checking if any other task finishes.

I've updated the task to loop through each job, waiting for it to finish. 

If a task fails, any task that is still running is stopped/aborted/cancelled and no output from them is returned. This is by design so that someone can go to the end of the built output to see what caused the job to fail.

I've also updated the task to support running multiple synchronous tasks in one background job.